README.md
# Github Webhook -> Kafka

This is a simple Github web hook server which streams incoming events into a Kafka queue.  Each incoming web hook event will be validated with the Github secret and any incoming requests not matching the validation will be rejected with a HTTP status code `401 Unauthorized`.

## Configuration

The following environment variables need to be set:

- `GWK_GITHUB_SECRET` - this is the secret used when creating / registering the web hook.  Defaults to empty string
- `GWK_KAFKA_BROKERS` - a comma separated list of Kafka servers. Defaults to `localhost:9092`
- `GWK_PORT` - webserver HTTP port. defaults to `8000`

Optionally, you can control debugging with:

- `GWK_KAFKA_TOPIC_PREFIX` - the kafka topic prefix which is append to the Github event name.  defaults to `github/unprocessed/`
- `GWK_PAYLOAD_DEBUG` - dump each incoming JSON event payload to the console when `true`
- `GWK_KAFKA_DEBUG` - log kafka details when `true`

## Running

You can easily run using Docker.

If you want to use a pre-built container ([`jhaynie/github-webhook-kafka`](https://hub.docker.com/r/jhaynie/github-webhook-kafka/)), you could use something like:

`docker run -d -p 8000:8000 -e GWK_GITHUB_SECRET=hello -e GWK_KAFKA_BROKERS=192.168.1.1:9092 jhaynie/github-webhook-kafka`

Make sure you use the correct `GWK_GITHUB_SECRET` and `GWK_KAFKA_BROKERS` values.  This will run the container on port 8000 by default.

For development, you can use the all-in-one container which contains Kafka, Zookeeper and this software:

`docker run -d -p 8000:8000 -p 2181:2181 -p 9092:9092 -e GWK_GITHUB_SECRET=hello -e GWK_ADVERTISED_HOST=192.168.1.1 jhaynie/github-webhook-kafka-dev`

Make sure you use the correct `GWK_GITHUB_SECRET`, `GWK_ADVERTISED_HOST` values.


## License

Copyright (c) 2016 by Jeff Haynie

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

## 1.0.0 - 09/28/2016

- initial release

## 1.0.1 - 09/29/2016

- parse incoming event payload and place the event into the queue instead of the entire payload
- changed `KAFKA_TOPIC` to `KAFKA_TOPIC_PREFIX` and prepend this string to the name of the event to form topic name
- switched to [landoop/fast-data-dev](https://hub.docker.com/r/landoop/fast-data-dev/) for dev container
- renamed debug log prefix from `sarama` to `github-webhook-kafka`
- renamed all environment variables with `GWK_` prefix

## 1.0.2

- set `config.Producer.MaxMessageBytes = 10000000`
- set `config.Producer.Compression = CompressionGZIP`
- in the dev container, set `message.max.bytes=10000000` for kafka broker for larger message bodies by default
- revert to use full event instead of event payload from previous release so additional metadata can be delivered
- changed format of kafka value to be JSON with string payload of body
- ability to specify an encrypted context in the url path which will be automatically decrypted with the provided key and provided in the kafka event payload
- provide an ability to generate a key from command line by passing two arguments (key as 16 characters and text to encrypt)
- switch to emit protobuf encoded messages using [go-github-protobuf](https://github.com/jhaynie/go-github-protobuf)


message.pb.go
// Code generated by protoc-gen-go.
// source: message.proto
// DO NOT EDIT!

/*
Package main is a generated protocol buffer package.

It is generated from these files:
	message.proto

It has these top-level messages:
	Message
*/
package main

import proto "github.com/golang/protobuf/proto"
import fmt "fmt"
import math "math"

// Reference imports to suppress errors if they are not otherwise used.
var _ = proto.Marshal
var _ = fmt.Errorf
var _ = math.Inf

// This is a compile-time assertion to ensure that this generated file
// is compatible with the proto package it is being compiled against.
// A compilation error at this line likely means your copy of the
// proto package needs to be updated.
const _ = proto.ProtoPackageIsVersion2 // please upgrade the proto package

type Message struct {
	Version   int32  `protobuf:"varint,1,opt,name=version" json:"version,omitempty"`
	Timestamp string `protobuf:"bytes,2,opt,name=timestamp" json:"timestamp,omitempty"`
	Id        string `protobuf:"bytes,3,opt,name=id" json:"id,omitempty"`
	Event     string `protobuf:"bytes,4,opt,name=event" json:"event,omitempty"`
	Payload   []byte `protobuf:"bytes,5,opt,name=payload,proto3" json:"payload,omitempty"`
	Context   string `protobuf:"bytes,6,opt,name=context" json:"context,omitempty"`
}

func (m *Message) Reset()                    { *m = Message{} }
func (m *Message) String() string            { return proto.CompactTextString(m) }
func (*Message) ProtoMessage()               {}
func (*Message) Descriptor() ([]byte, []int) { return fileDescriptor0, []int{0} }

func init() {
	proto.RegisterType((*Message)(nil), "main.Message")
}

func init() { proto.RegisterFile("message.proto", fileDescriptor0) }

var fileDescriptor0 = []byte{
	// 155 bytes of a gzipped FileDescriptorProto
	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x09, 0x6e, 0x88, 0x02, 0xff, 0xe2, 0xe2, 0xcd, 0x4d, 0x2d, 0x2e,
	0x4e, 0x4c, 0x4f, 0xd5, 0x2b, 0x28, 0xca, 0x2f, 0xc9, 0x17, 0x62, 0xc9, 0x4d, 0xcc, 0xcc, 0x53,
	0x9a, 0xcd, 0xc8, 0xc5, 0xee, 0x0b, 0x11, 0x17, 0x92, 0xe0, 0x62, 0x2f, 0x4b, 0x2d, 0x2a, 0xce,
	0xcc, 0xcf, 0x93, 0x60, 0x54, 0x60, 0xd4, 0x60, 0x0d, 0x82, 0x71, 0x85, 0x64, 0xb8, 0x38, 0x4b,
	0x32, 0x81, 0xda, 0x4b, 0x12, 0x73, 0x0b, 0x24, 0x98, 0x80, 0x72, 0x9c, 0x41, 0x08, 0x01, 0x21,
	0x3e, 0x2e, 0xa6, 0xcc, 0x14, 0x09, 0x66, 0xb0, 0x30, 0x90, 0x25, 0x24, 0xc2, 0xc5, 0x9a, 0x5a,
	0x96, 0x9a, 0x57, 0x22, 0xc1, 0x02, 0x16, 0x82, 0x70, 0x40, 0xa6, 0x17, 0x24, 0x56, 0xe6, 0xe4,
	0x27, 0xa6, 0x48, 0xb0, 0x02, 0xc5, 0x79, 0x82, 0x60, 0x5c, 0x90, 0x4c, 0x72, 0x7e, 0x5e, 0x49,
	0x6a, 0x45, 0x89, 0x04, 0x1b, 0x58, 0x07, 0x8c, 0x9b, 0xc4, 0x06, 0x76, 0xaa, 0x31, 0x20, 0x00,
	0x00, 0xff, 0xff, 0x71, 0x92, 0xe0, 0xb3, 0xbb, 0x00, 0x00, 0x00,
}


start-kafka.sh
#!/bin/bash

if [[ -z "$KAFKA_PORT" ]]; then
    export KAFKA_PORT=9092
fi
# if [[ -z "$KAFKA_ADVERTISED_PORT" ]]; then
#     export KAFKA_ADVERTISED_PORT=$(docker port `hostname` $KAFKA_PORT | sed -r "s/.*:(.*)/\1/g")
# fi
if [[ -z "$KAFKA_BROKER_ID" ]]; then
    # By default auto allocate broker ID
    export KAFKA_BROKER_ID=-1
fi
if [[ -z "$KAFKA_LOG_DIRS" ]]; then
    export KAFKA_LOG_DIRS="/kafka/kafka-logs-$HOSTNAME"
fi
# if [[ -z "$KAFKA_ZOOKEEPER_CONNECT" ]]; then
#     export KAFKA_ZOOKEEPER_CONNECT=$(env | grep ZK.*PORT_2181_TCP= | sed -e 's|.*tcp://||' | paste -sd ,)
# fi

if [[ -n "$KAFKA_HEAP_OPTS" ]]; then
    sed -r -i "s/(export KAFKA_HEAP_OPTS)=\"(.*)\"/\1=\"$KAFKA_HEAP_OPTS\"/g" $KAFKA_HOME/bin/kafka-server-start.sh
    unset KAFKA_HEAP_OPTS
fi

if [[ ! -z "$ADV_HOST" ]]; then
	export KAFKA_ADVERTISED_HOST_NAME=$ADV_HOST
fi

if [[ -z "$KAFKA_ADVERTISED_HOST_NAME" && -n "$HOSTNAME_COMMAND" ]]; then
    export KAFKA_ADVERTISED_HOST_NAME=$(eval $HOSTNAME_COMMAND)
fi

# Allow specification of log retention policies
if [ ! -z "$LOG_RETENTION_HOURS" ]; then
    echo "log retention hours: $LOG_RETENTION_HOURS"
    sed -r -i "s/(log.retention.hours)=(.*)/\1=$LOG_RETENTION_HOURS/g" $KAFKA_HOME/config/server.properties
fi

if [ ! -z "$LOG_RETENTION_BYTES" ]; then
    echo "log retention bytes: $LOG_RETENTION_BYTES"
    sed -r -i "s/#(log.retention.bytes)=(.*)/\1=$LOG_RETENTION_BYTES/g" $KAFKA_HOME/config/server.properties
fi

# Enable/disable auto creation of topics
if [ ! -z "$AUTO_CREATE_TOPICS" ]; then
    echo "auto.create.topics.enable: $AUTO_CREATE_TOPICS"
    echo "auto.create.topics.enable=$AUTO_CREATE_TOPICS" >> $KAFKA_HOME/config/server.properties
fi

if [ ! -z "$KAFKA_STANDALONE" ]; then
	echo "offsets.topic.replication.factor=1" >> $KAFKA_HOME/config/server.properties
fi

for VAR in `env`
do
  if [[ $VAR =~ ^KAFKA_ && ! $VAR =~ ^KAFKA_HOME ]]; then
    kafka_name=`echo "$VAR" | sed -r "s/KAFKA_(.*)=.*/\1/g" | tr '[:upper:]' '[:lower:]' | tr _ .`
    env_var=`echo "$VAR" | sed -r "s/(.*)=.*/\1/g"`
    if egrep -q "(^|^#)$kafka_name=" $KAFKA_HOME/config/server.properties; then
        sed -r -i "s@(^|^#)($kafka_name)=(.*)@\2=${!env_var}@g" $KAFKA_HOME/config/server.properties #note that no config values may contain an '@' char
    else
        echo "$kafka_name=${!env_var}" >> $KAFKA_HOME/config/server.properties
    fi
  fi
done

# Generate the config only if it doesn't exist
if [ ! -f "$ZOO_CONF_DIR/zoo.cfg" ]; then
    CONFIG="$ZOO_CONF_DIR/zoo.cfg"

    echo "clientPort=$ZOO_PORT" >> "$CONFIG"
    echo "dataDir=$ZOO_DATA_DIR" >> "$CONFIG"
    echo "dataLogDir=$ZOO_DATA_LOG_DIR" >> "$CONFIG"

	 if [ ! -z "$KAFKA_ADVERTISED_HOST_NAME" ]; then
		 echo "advertised.host.name=$KAFKA_ADVERTISED_HOST_NAME"  >> "$CONFIG"
	 fi

    echo 'tickTime=2000' >> "$CONFIG"
    echo 'initLimit=5' >> "$CONFIG"
    echo 'syncLimit=2' >> "$CONFIG"

    for server in $ZOO_SERVERS; do
        echo "$server" >> "$CONFIG"
    done

fi

# Write myid only if it doesn't exist
if [ ! -f "$ZOO_DATA_DIR/myid" ]; then
    echo "${ZOO_MY_ID:-1}" > "$ZOO_DATA_DIR/myid"
fi

if [[ -n "$CUSTOM_INIT_SCRIPT" ]] ; then
  eval $CUSTOM_INIT_SCRIPT
fi


KAFKA_PID=0

# see https://medium.com/@gchudnov/trapping-signals-in-docker-containers-7a57fdda7d86#.bh35ir4u5
term_handler() {
  echo 'Stopping Kafka....'
  if [ $KAFKA_PID -ne 0 ]; then
    kill -s TERM "$KAFKA_PID"
    wait "$KAFKA_PID"
  fi
  echo 'Kafka stopped.'
  exit
}


# Capture kill requests to stop properly
trap "term_handler" SIGHUP SIGINT SIGTERM

supervisord -n &
KAFKA_PID=$!

wait


main.go
package main

import (
	"fmt"
	"os"
	"io"
	"log"
	"time"
	"errors"
	"bytes"
	"strings"
	"os/exec"
	"os/signal"
	"net/http"
	"crypto/rand"
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"encoding/json"

	"github.com/jhaynie/go-github-protobuf/github"
	"github.com/golang/protobuf/proto"
	"github.com/satori/go.uuid"
	"github.com/Shopify/sarama"
)

var Version string
var Build string
var Key string
var Producer sarama.SyncProducer

func getenv(key string, def string) string {
	val := os.Getenv(key)
	if val == "" {
		return def
	}
	return val
}

func stringify(v interface{}) string {
	buf, err := json.MarshalIndent(v, "", "\t")
	if err != nil {
		return ""
	}
	return string(buf)
}


func decrypt(text []byte) (string, error) {
	block, err := aes.NewCipher([]byte(Key))
	if err != nil {
		return "", err
	}
	if len(text) < aes.BlockSize {
		return "", errors.New("ciphertext too short")
	}
	iv := text[:aes.BlockSize]
	text = text[aes.BlockSize:]
	cfb := cipher.NewCFBDecrypter(block, iv)
	cfb.XORKeyStream(text, text)
	data, err := base64.StdEncoding.DecodeString(string(text))
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func encrypt(text []byte) (string, error) {
	block, err := aes.NewCipher([]byte(Key))
	if err != nil {
		return "", err
	}
	b := base64.StdEncoding.EncodeToString(text)
	ciphertext := make([]byte, aes.BlockSize+len(b))
	iv := ciphertext[:aes.BlockSize]
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}
	cfb := cipher.NewCFBEncrypter(block, iv)
	cfb.XORKeyStream(ciphertext[aes.BlockSize:], []byte(b))
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func handler(w http.ResponseWriter, r *http.Request) {
	var context string
	if Key != "" {
		// if we have a configured Key and we are passed in a context,
		// we are going to decrypt it with the provided Key
		buf, err := base64.StdEncoding.DecodeString(r.URL.Path[1:])
		if err != nil {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}
		dec, err := decrypt(buf)
		if err != nil {
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}
		context = dec
	}
	payload, err := github.ValidatePayload(r, []byte(getenv("GWK_GITHUB_SECRET","")))
	if err != nil {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
	} else {
		name := github.WebHookType(r)
		log.Println(string(payload))
		event, err := github.ParseWebHook(name, payload)
		if err != nil {
			log.Println(err)
			http.Error(w, "error handling message", http.StatusInternalServerError)
		} else {
			protobuf, err := proto.Marshal(event.(proto.Message))
			if err != nil {
				log.Println(err)
				http.Error(w, "error handling message", http.StatusInternalServerError)
			} else {
				message := Message{
					Version: 1,
					Timestamp: time.Now().UTC().Format(time.RFC3339),
					Id: uuid.NewV4().String(),
					Event: name,
					Payload: protobuf,
					Context: context,
				}
				log.Println(stringify(message))
				value, err := proto.Marshal(&message)
				if err != nil {
					log.Println(err)
					http.Error(w, "error handling message", http.StatusInternalServerError)
				} else {
					topic := fmt.Sprintf("%s%s", getenv("GWK_KAFKA_TOPIC_PREFIX",""), name)
					if getenv("GWK_PAYLOAD_DEBUG", "") == "true" {
						log.Println(fmt.Sprintf("sending to %s", topic))
					}
					_, _, err := Producer.SendMessage(&sarama.ProducerMessage{
						Topic: topic,
						Value: sarama.ByteEncoder(value),
					})
					if err != nil {
						log.Println(err)
						http.Error(w, "error handling message", http.StatusInternalServerError)
					}
					w.WriteHeader(http.StatusAccepted)
				}
			}
		}
	}
}

func createProducer() sarama.SyncProducer {
	if getenv("GWK_KAFKA_DEBUG", "") == "true" {
		sarama.Logger = log.New(os.Stdout, "[github-webhook-kafka] ", log.LstdFlags)
	}
	brokerList := strings.Split(getenv("GWK_KAFKA_BROKERS","localhost:9092"), ",")
	config := sarama.NewConfig()
	config.Producer.Compression = sarama.CompressionGZIP
	config.Producer.MaxMessageBytes = 10000000
	config.Producer.RequiredAcks = sarama.WaitForAll
	config.Producer.Retry.Max = 100
	producer, err := sarama.NewSyncProducer(brokerList, config)
	if err != nil {
		log.Fatal(err)
	}
	return producer
}

func main() {
	if len(os.Args) > 2 {
		Key = os.Args[1]
		buf, err := encrypt([]byte(os.Args[2]))
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(buf)
		os.Exit(0)
	}
	Key = getenv("GWK_AES_KEY","")
	Producer = createProducer()
	mux := http.NewServeMux()
 	mux.HandleFunc("/", handler)
	listen := fmt.Sprintf(":%s", getenv("GWK_PORT","8000"))
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	go func(){
		for range c {
			log.Println("[github-webhook-kafka] SIGINT, exiting...")
			Producer.Close()
			os.Exit(0)
		}
	}()
	go func(){
		cmd := exec.Command("lt", "--port", "8000")
		var out bytes.Buffer
		cmd.Stdout = &out
		err := cmd.Start()
		if err != nil {
			log.Fatal(err)
		}
		for i := 0; i < 10; i++ {
			s := strings.TrimSpace(out.String())
			if s == "" {
				time.Sleep(1 * time.Second)
				continue
			}
	    	fmt.Printf("[github-webhook-kafka] URL: \033[31m%s\033[m\n", strings.Replace(s, "your url is:", "", -1))
			break
		}
	}()
	log.Printf("[github-webhook-kafka] listening on port %s\n", listen[1:])
	err := http.ListenAndServe(listen, mux)
	if err != nil {
		log.Fatal(err)
	}
}
